File: Tree/66.Binary Tree Preorder Traversal/Solution.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

# Non-Recursion (Stack)

class Solution:
    """
    @param root: A Tree
    @return: Preorder in ArrayList which contains node values.
    """
    def preorderTraversal(self, root):
        # write your code here
        stack  = []
        result = []

        if root is None:
            return result

        stack.append(root)

        # while stack is not Empty:
        while stack:
            node = stack.pop()
            result.append(node.val)

            # stack.push(node.right)
            if node.right is not None:
                stack.append(node.right)

            # stack.push(node.left)
            if node.left is not None:
                stack.append(node.left)

        return result



"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

# Traverse

class Solution:
    """
    @param root: A Tree
    @return: Preorder in ArrayList which contains node values.
    """
    def preorderTraversal(self, root):
        # write your code here
        result = []
        self.helper(root, result)
        return result


    def helper(self, root, result):
        if root is None:
            return

        result.append(root.val)
        self.helper(root.left, result)
        self.helper(root.right, result)



"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

# Divide & Conquer

class Solution:
    """
    @param root: A Tree
    @return: Preorder in ArrayList which contains node values.
    """
    def preorderTraversal(self, root):
        # write your code here
        result = []
        if root is None:
            return result

        left  = self.preorderTraversal(root.left)
        right = self.preorderTraversal(root.right)

        result.append(root.val)
        result.extend(left)
        result.extend(right)

        return result

File: Tree/66.Binary Tree Preorder Traversal/Solution.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */


// Solution 1: Traverse helper function - Template 1

public class Solution {
    /**
     * @param root: A Tree
     * @return: Preorder in ArrayList which contains node values.
     */
    public List<Integer> preorderTraversal(TreeNode root) {
        // write your code here
        List<Integer> result = new ArrayList<>();
        traverse(root, result);
        return result;

    }

    private void traverse(TreeNode root, List<Integer> result) {
        // return if current root is null.
        if (root == null) {
            return;
        }

        // 根 -> 左 -> 右
        result.add(root.val);
        traverse(root.left, result);
        traverse(root.right, result);
    }

}


// Solution 2: Dicide and Conquer - Template 2

public class Solution {
    /**
     * @param root: A Tree
     * @return: Preorder in ArrayList which contains node values.
     */
    public List<Integer> preorderTraversal(TreeNode root) {
        // write your code here
        // Divide and Conquer
        List<Integer> result = new ArrayList<Integer> ();
        if (root == null) {
            return result;
        }

        List<Integer> left = preorderTraversal(root.left);
        List<Integer> right = preorderTraversal(root.right);

        result.add(root.val);
        result.addAll(left);
        result.addAll(right);

        return result;
        
    }

}


// Solution 3: Non-Recursive (Stack)

public class Solution {
    /**
     * @param root: A Tree
     * @return: Preorder in ArrayList which contains node values.
     */
    public List<Integer> preorderTraversal(TreeNode root) {
        // write your code here
        // Divide and Conquer
        List<Integer> result = new ArrayList<Integer> ();
        Stack<TreeNode> NodeStack = new Stack<TreeNode> ();
        
        if (root == null) {
            return result;
        }

        NodeStack.push(root);
        
        while (!NodeStack.isEmpty()) {

            TreeNode node = NodeStack.pop();
            result.add(node.val);

            if (node.right != null) NodeStack.push(root.right);
            if (node.left != null) NodeStack.push(root.left);

        }

        return result;
        
    }

}